Support incremental in compiletest for non-incremental modes.

ehuss opened the pull request:
This adds first-class support for using incremental builds in non-incremental-mode tests.  These tests previously manually passed `-C incremental=tmp/foo` which resulted in reusing the same tmp folder between runs.  This means that these tests could fail whenever the on-disk incremental format changed (such as when updating one's local source tree).  This changes it so that these tests can pass a `// incremental-build` header which instructs compiletest to create a set aside a dedicated incremental directory which will be cleared before the test starts to ensure it has a clean slate.

rust-highfive commented:
r? @Mark-Simulacrum

(rust-highfive has picked a reviewer for you, use r? to override)

ehuss commented:
There was prior discussion about this at https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Incremental.20ICE.20running.20rustc.20tests.20on.20origin.2Fmaster/near/251211217

It was suggested that these tests could be moved to the incremental directory, but I don't think that is possible because these are codegen tests which are tested differently.  I'm not familiar with these tests, so maybe they can be moved, so let me know if an alternate solution would be preferred.

cc @michaelwoerister

Mark-Simulacrum commented on src/tools/compiletest/src/header.rs:
```diff
@@ -113,6 +113,8 @@ pub struct TestProps {
     // testing harness and used when generating compilation
     // arguments. (In particular, it propagates to the aux-builds.)
     pub incremental_dir: Option<PathBuf>,
+    // If `true`, this test will use incremental compilation.
```
Can we add some docs here (since this is basically where this is documented) about what 'scenario' of incr comp the test will be run under? In particular, I think we will ensure that the test runs each time starting with a clean incr. comp. dir and then any revisions will continue in the same directory (i.e., become "incr-patched" scenarios effectively).

ehuss replied:
Of course!  I was intending to update the [rustc dev guide docs](https://rustc-dev-guide.rust-lang.org/tests/adding.html#other-header-commands) after this merged, but I understand that some people will only read one source or the other.

Mark-Simulacrum commented:
I think we should also consider renaming the flag to "incremental" or "incremental-mode" -- I want to avoid accidentally reading it as similar to 'build-pass' or otherwise tying it specifically to *build* in that sense.

r=me with the docs and a possible rename, happy to leave it up to you which to choose -- or if you feel strongly about -build, that's also OK.

ehuss commented:
Yea, I can see how that can be confusing with `-build`.  I originally used just `incremental`, but then I noticed that a single English word can be accidentally picked up by a comment like [this](https://github.com/rust-lang/rust/blob/bf642323d621dcefeef1d8ab4711aae36e357615/src/test/codegen-units/partitioning/extern-generic.rs#L2-L3).

I went ahead and renamed it back to just `incremental`, since I think `incremental-mode` would be confusing with true incremental mode tests that support passes. I think the risk of accidentally making a test incremental is low (the existing examples all use `-C incremental`), and the consequences are low (making a test incremental usually shouldn't interfere with it).

I kinda wish the header parsing was a little more strict so that risk of accidental matches wasn't there.

Feel free to r- if that doesn't sound good, or if the comment I added isn't correct or clear.

ehuss commented:
@bors r=Mark-Simulacrum

bors commented:
:pushpin: Commit 2da6e669488791bd2d36e3a35fa2e4e4e02bc647 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 2da6e669488791bd2d36e3a35fa2e4e4e02bc647 -->
<!-- homu: {"type":"Approved","sha":"2da6e669488791bd2d36e3a35fa2e4e4e02bc647","approver":"Mark-Simulacrum"} -->

Mark-Simulacrum commented:
Sounds great, thanks! Yeah, our header parsing story is not great :)

bors commented:
:hourglass: Testing commit 2da6e669488791bd2d36e3a35fa2e4e4e02bc647 with merge ac8dd1b2f24dc62c962172b27433106b4e84dc62...
<!-- homu: {"type":"BuildStarted","head_sha":"2da6e669488791bd2d36e3a35fa2e4e4e02bc647","merge_sha":"ac8dd1b2f24dc62c962172b27433106b4e84dc62"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3712521593)
Approved by: Mark-Simulacrum
Pushing ac8dd1b2f24dc62c962172b27433106b4e84dc62 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/3712521593"},"merge_sha":"ac8dd1b2f24dc62c962172b27433106b4e84dc62"} -->

rust-timer commented:
Finished benchmarking commit (ac8dd1b2f24dc62c962172b27433106b4e84dc62): [comparison url](https://perf.rust-lang.org/compare.html?start=f6e6ddc09d660fbfb4d483a9677d80768cc4e31c&end=ac8dd1b2f24dc62c962172b27433106b4e84dc62).

**Summary**: This benchmark run did not return any relevant changes.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

 

@rustbot label: -perf-regression
